scripts/count_decls.lean

```diff
@@ -0,0 +1,86 @@
+/-
+Copyright (c) 2024 Kyle Miller. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kyle Miller, Damiano Testa
+-/
+import Mathlib
+
+/-!
+# `count_decls` -- a tally of declarations in `Mathlib`
+
+This file is used by the `Periodic reports` script to generate a tally of declarations
+in `Mathlib`, `Batteries` and core.
+
+Source: https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/Metamathematics.3A.20Theorems.20by.20Domain.20Areas/near/374090639
+-/
+
+namespace PeriodicReports
+open Lean Elab Meta
+
+/-- A structure to tally declaration types:
+* `thms` are theorems;
+* `preds` are predicates;
+* `types` are types;
+* `data` is everything else.
+-/
+structure Tally where
+  thms  : HashSet Name := {}
+  data  : HashSet Name := {}
+  preds : HashSet Name := {}
+  types : HashSet Name := {}
+
+/-- `toString t` produces a string where each line is either
+* the (human-readable) name of a field of `t : Tally`, or
+* the name of a single declaration followed by `,`.
+
+The script to extract comparison data relies on the format
+* lines ending with `,` or not,
+* single declaration per line.
+-/
+def toString (t : Tally) : String :=
+  let print (h : HashSet Name) : String :=
+    let tot := h.toList.map (·.toString)
+    String.intercalate ",\n" tot ++ ","
+  s!"\
+Theorems
+{print t.thms}
+Data
+{print t.data}
+Predicates
+{print t.preds}
+Types
+{print t.types}"
+
+variable (mods : HashSet Nat) (env : Environment) in
+/-- Extend a `Tally` by the ConstantInfo `c`.  It is written to work with `Lean.SMap.foldM`. -/
+def updateTally (s : Tally) (n : Name) (c : ConstantInfo) :
+    MetaM Tally := do
+  if c.isUnsafe || (← n.isBlackListed) then return s else
+  let typ := c.type
+  if (← isProp typ) then
+    return {s with thms := s.thms.insert  n}
+  else
+    let exp ← forallTelescopeReducing typ fun _ e => return e
+    if exp.isType then
+      return {s with types := s.types.insert n}
+    else
+    if exp.isSort then
+      return {s with preds := s.preds.insert n}
+    else
+      return {s with data := s.data.insert n}
```

Comment:
Maybe for clarity (and future extension) it would be better to nest inside the `forallTelescopeReducing`?
```suggestion
    forallTelescopeReducing typ fun _ exp' => do
      if exp'.isType then
        return {s with types := s.types.insert n}
      else
      if exp'.isSort then
        return {s with preds := s.preds.insert n}
      else
        return {s with data := s.data.insert n}
```